framework/topology.cpp

```diff
@@ -656,19 +534,65 @@ static bool fill_ppin_sysfs(struct cpu_info *info)
     return false;
 }
 
+static unsigned  get_apic_id_from_cpuid() {
+    #ifdef _WIN32
+        unsigned int cpu_info[4];
+        __cpuid(cpu_info, 1);
+        return (cpu_info[1] >> 24) & 0xFF;
+    #else
+        unsigned eax, ebx, ecx, edx;
+        __cpuid(1, eax, ebx, ecx, edx);
+        return (ebx >> 24) & 0xFF;
+    #endif
+    }
+
 static bool fill_topology_hwloc(struct cpu_info *info, hwloc_topology_t topology)
 {
-    hwloc_obj_t obj = hwloc_get_obj_by_type(topology, HWLOC_OBJ_PU, info->cpu_number);
-    if (!obj) {
+    hwloc_obj_t pu = hwloc_get_pu_obj_by_os_index(topology, info->cpu_number);
+    if (!pu) {
         hwloc_topology_destroy(topology);
         return false;
     }
 
-    info->package_id = hwloc_get_ancestor_obj_by_type(topology, HWLOC_OBJ_PACKAGE, obj)->logical_index;
-    info->core_id = hwloc_get_ancestor_obj_by_type(topology, HWLOC_OBJ_CORE, obj)->logical_index;
-    info->thread_id = obj->logical_index;
-    info->apic_id = obj->os_index;
+    int apic = 0;
 
+    // Try this 1st to find APIC ID in infos[]
+    for (unsigned i = 0; i < pu->infos_count; i++) {
+        if (strcmp(pu->infos[i].name, "APIC ID") == 0) {
+            info->apic_id = atoi(pu->infos[i].value);
+            apic = 1;
+            break;
+        }
+    }
+
+    // If the above lookup fails execute the below
+    if (!apic) {
+        int can_bind = hwloc_topology_get_support(topology)->cpubind->set_thisproc_cpubind;
+        if (can_bind) {
+            hwloc_cpuset_t cpuset = hwloc_bitmap_dup(pu->cpuset);
+            hwloc_bitmap_singlify(cpuset);
+            if (hwloc_set_cpubind(topology, cpuset, HWLOC_CPUBIND_PROCESS) == 0) {
+                info->apic_id = get_apic_id_from_cpuid();
+                apic = 1;
+            }
+            hwloc_bitmap_free(cpuset);
+        }
```

Comment:
We're already bound to the right PU. But I'd rather you left the APIC ID set to -1 here and added another function to fill the missing bits in.